gateway-server/src/main/java/org/apache/knox/gateway/webshell/WebShellTokenUtils.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.knox.gateway.services.security.token.UnknownTokenException;
+import org.apache.knox.gateway.services.security.token.impl.JWT;
+import org.apache.knox.gateway.services.security.token.impl.JWTToken;
+import org.apache.knox.gateway.services.topology.TopologyService;
+import org.apache.knox.gateway.topology.Service;
+import org.apache.knox.gateway.topology.Topology;
+import org.apache.knox.gateway.util.CertificateUtils;
+import org.apache.knox.gateway.util.Tokens;
+import org.eclipse.jetty.websocket.servlet.ServletUpgradeRequest;
+import java.net.HttpCookie;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+public class WebShellTokenUtils {
+    private static final String KNOXSSO_COOKIE_NAME = "knoxsso.cookie.name";
+    private static final String DEFAULT_SSO_COOKIE_NAME = "hadoop-jwt";
+    private static final String JWT_DEFAULT_ISSUER = "KNOXSSO";
+    private static final String JWT_EXPECTED_ISSUER = "jwt.expected.issuer";
+    private static final String JWT_DEFAULT_SIGALG = "RS256";
+    private static final String JWT_EXPECTED_SIGALG = "jwt.expected.sigalg";
+    private static final String SSO_VERIFICATION_PEM = "sso.token.verification.pem";
+
+    private static JWTMessages log = MessagesFactory.get(JWTMessages.class);
+    private static String cookieName;
+    private static String expectedIssuer;
+    private static String expectedSigAlg;
+    private static TokenStateService tokenStateService;
+    private static JWTokenAuthority authorityService;
+    private static RSAPublicKey publicKey;
+    // todo: how to initialize these two attributes.
+    private static SignatureVerificationCache signatureVerificationCache;
+    private static String expectedJWKSUrl;
+
+    public static void validateJWT(ServletUpgradeRequest req, GatewayServices services, GatewayConfig gatewayConfig) throws UnknownTokenException {
+        configureParameters(services, gatewayConfig);
+        List<HttpCookie> ssoCookies = req.getCookies();
+        for (HttpCookie ssoCookie : ssoCookies) {
+            if (cookieName.equals(ssoCookie.getName())) {
+                try {
+                    JWT token = new JWTToken(ssoCookie.getValue());
+                    if (validateToken(token)) {
+                        // we found a valid cookie we don't need to keep checking anymore
+                        return;
+                    }
+                } catch (ParseException | UnknownTokenException ignore) {
+                    // Ignore the error since cookie was invalid
+                    // Fall through to keep checking if there are more cookies
+                }
+            }
+        }
+        // no cookie contain a valid JWT
+        throw new UnknownTokenException("No valid JWT found for webshell");
+    }
+
+    private static void configureParameters(GatewayServices services, GatewayConfig gatewayConfig) {
```

Comment:
We should check if parameters are already configured, if they are then no need to call this again. One thing to note is that we need to check if topology has reloaded, we can check that using `topology.getTimestamp()` method